### ISSUE TITLE
about FedProx

### ISSUE THREAD

**Issue by KevinM1ao** [woodenchild95/FL-Simulator#12]
Hi, I have a bit of a question about the implementation part of your fedprox algorithm, could you please answer it?

![1713166293159](https://github.com/woodenchild95/FL-Simulator/assets/130957078/375b1a20-2aed-46e3-a715-d094948076b7)
![1713166501723](https://github.com/woodenchild95/FL-Simulator/assets/130957078/d074ef8f-2c8a-4851-8196-9a065785d5aa)

In the paper, the loss function is followed by an L2 regularization term for the current model and the global model, but it seems that their L2 regularization is not computed in your code:torch.sum(param_list * delta_list) * -1.

**Comment by woodenchild95** [comment]
> Hi, I have a bit of a question about the implementation part of your fedprox algorithm, could you please answer it?
> 
> ![1713166293159](https://private-user-images.githubusercontent.com/130957078/322391477-375b1a20-2aed-46e3-a715-d094948076b7.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMxNjc3MjUsIm5iZiI6MTcxMzE2NzQyNSwicGF0aCI6Ii8xMzA5NTcwNzgvMzIyMzkxNDc3LTM3NWIxYTIwLTJhZWQtNDZlMy1hNzE1LWQwOTQ5NDgwNzZiNy5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxNVQwNzUwMjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYzViZjc3MTM0Yjg5OWEzNjI4MDgzOWQ4ZWJhZjFhNzI4MjllYjdmMDkyODkwOTBiMGZmMmMzZThhNjE0YzZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.kEeJJW8WGNGvuhtumvqkoEGjl21p-I-t9quqyghD-24) ![1713166501723](https://private-user-images.githubusercontent.com/130957078/322391763-d074ef8f-2c8a-4851-8196-9a065785d5aa.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMxNjc3MjUsIm5iZiI6MTcxMzE2NzQyNSwicGF0aCI6Ii8xMzA5NTcwNzgvMzIyMzkxNzYzLWQwNzRlZjhmLTJjOGEtNDg1MS04MTk2LTlhMDY1Nzg1ZDVhYS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxNVQwNzUwMjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lN2ZmZGM4NGFkMTgwMzdmZTk3ZjcwMWFjZGJjN2VkYjU5YzM2MDQ4ZTg2OGY0M2Y1MTU1OTczZGMyNmM0YmNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.O2SL0f877NYrOOC-0N-9jHRCeLo-gCyf2dTdSRLBIKU)
> 
> In the paper, the loss function is followed by an L2 regularization term for the current model and the global model, but it seems that their L2 regularization is not computed in your code:torch.sum(param_list * delta_list) * -1.

Hi, @KevinM1ao, as the problem is defined: $\min_w h = f(w) + \frac{\mu}{2}||w - w^t||^2 = \min_w f(w) + \frac{\mu}{2}||w||^2 + \mu\langle w, w^t \rangle$. After reconstruction, $\frac{\mu}{2}||w||^2$ is the regularization loss and $\mu\langle w, w^t \rangle$ is the loss in the red blank above. The first loss is added to the weight decay in the local optimizer, and the second loss is constructed above. You can also implement a new optimizer class to perform the $w = w - \eta(g + \mu(w-w^t))$. They are equal.

**Comment by KevinM1ao** [comment]
thanks so much!
